src/sys/fuchsia/selector.rs

```diff
@@ -0,0 +1,355 @@
+use {io, Event, PollOpt, Ready, Token};
+use sys::fuchsia::{
+    epoll_event_to_ready,
+    poll_opts_to_wait_async,
+    status_to_io_err,
+    EventedFd,
+    EventedFdInner,
+};
+use concurrent_hashmap::ConcHashMap;
+use magenta;
+use magenta_sys;
+use magenta::HandleBase;
+use std::collections::hash_map::RandomState;
+use std::fmt;
+use std::mem;
+use std::sync::atomic::{AtomicBool, AtomicUsize, ATOMIC_USIZE_INIT, Ordering};
+use std::sync::{Arc, Mutex, Weak};
+use std::time::Duration;
+use sys;
+
+/// The kind of registration-- file descriptor or handle.
+///
+/// The last bit of a token is set to indicate the type of the registration.
+#[derive(Copy, Clone, Eq, PartialEq)]
+enum RegType {
+    Fd,
+    Handle,
+}
+
+fn key_from_token_and_type(token: Token, reg_type: RegType) -> u64 {
+    let key = token.0 as u64;
+    let msb = 1u64 << 63;
+    if (key & msb) != 0 {
+        panic!("Most-significant bit of token must remain unset.");
```

Comment:
Could this return an error instead of panicking? It's sort of unfortunate that a bit has to be reserved b/c `Token` in general is intended to be a bag of bits for everyone else, but it seems the implementation currently necessitates this restrictoin.